tests/testsuite/registry_overlay.rs

```diff
@@ -0,0 +1,161 @@
+//! Tests for local-registry sources.
+
+use cargo_test_support::registry::{Package, RegistryBuilder, TestRegistry};
+use cargo_test_support::{project, Project};
+
+fn setup() -> (TestRegistry, String) {
+    let alt = RegistryBuilder::new().alternative().build();
+    (
+        RegistryBuilder::new().http_index().build(),
+        alt.index_url()
+            .to_file_path()
+            .unwrap()
+            .into_os_string()
+            .into_string()
+            .unwrap(),
+    )
+}
+
+fn simple_project() -> Project {
+    project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                bar = ">= 0.0.0"
+                baz = ">= 0.1.0"
+            "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .build()
+}
+
+#[cargo_test]
+fn overlay_hit() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    // baz is only in the local registry, but it gets found
+    Package::new("baz", "0.1.1")
+        .alternative(true)
+        .local(true)
+        .publish();
+
+    p.cargo("check")
+        .overlay_registry(&reg.index_url(), &alt_path)
+        .run();
+}
+
+#[cargo_test]
+fn version_fallback() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    // The one we want is in the main registry.
+    Package::new("baz", "0.1.1").publish();
+    // An older version in the local registry doesn't confuse it.
+    Package::new("baz", "0.0.1")
+        .alternative(true)
+        .local(true)
+        .publish();
```

Comment:
I don't think this is really testing anything because `baz` is too low of a version for the version requirement

Reply:
The motivation here was to test that the overlay is correctly the union of the versions in the overlay and the overlaid registries. (If it were only returning the overlay versions, it would only return the too-old version and fail to resolve)

Reply:
Except we aren't checking the resolve error for whether it sees `baz@0.0.1`

We also should probably have a test the other direction, that the overlay source properly shadows the regular source.

Reply:
Got it. I'm now testing both directions, and checking the appropriate versions in the stderr.